zopkio/adhoc_deployer.py

```diff
@@ -135,10 +135,37 @@ def install(self, unique_id, configs=None):
         if executable is None:
           logger.error("executable was not provided for unique_id: " + unique_id)
           raise DeploymentError("executable was not provided for unique_id: " + unique_id)
+
+        #if the executable is in remote location copy to local machine
+        copy_from_remote_location = False;
+        if (":" in executable):
+          copy_from_remote_location = True
+          remote_location_server = executable.split(":")[0]
+          remote_file_path = executable.split(":")[1] 
+          remote_file_name = os.path.basename(remote_file_path)
+
+          if (configs.get("temp_scratch")):
+            local_temp_file_name = configs["temp_scratch"] + "/" + remote_file_name
+          else:
+            local_temp_file_name = "/tmp/" + remote_file_name
+          
+          with get_sftp_client(remote_location_server,username=runtime.get_username(), password=runtime.get_password()) as ftp:
+            try:
+              ftp.get(remote_file_path, local_temp_file_name)
+              executable = local_temp_file_name
+            except:
+              raise DeploymentError("Unable to load file from remote server " + executable)
+                     
         exec_name = os.path.basename(executable)
         install_location = os.path.join(install_path, exec_name)
         with get_sftp_client(hostname, username=runtime.get_username(), password=runtime.get_password()) as ftp:
-          ftp.put(executable, install_location)
+          try:
+            ftp.put(executable, install_location)
+          except:
+            raise DeploymentError("Unable to copy executable to install_location:" + install_location)
+          finally:
```

Comment:
this finally block should be attached to a try starting at line 158 so it always runs if the file is downloaded